src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java

```diff
@@ -141,9 +167,62 @@ private void changeReaderConnectionTo(JdbcConnection newConnection) throws SQLEx
 
   @Override
   public void openInitialConnection(ConnectionUrl connectionUrl) throws SQLException {
-    this.hosts.addAll(connectionUrl.getHostsList());
     this.nextPlugin.openInitialConnection(connectionUrl);
-    this.writerConnection = this.currentConnectionProvider.getCurrentConnection();
+    final JdbcConnection currentConnection = this.currentConnectionProvider.getCurrentConnection();
+    if (currentConnection == null) {
+      return;
+    }
+
+    if (MULTI_HOST_CONNECTION_AWS.equals(connectionUrl.getType())) {
+      this.hosts.addAll(connectionUrl.getHostsList());
+      final HostInfo currentHost = this.rdsHostUtils.getHostInfo(this.hosts, currentConnection.getHostPortPair());
+      updateInternalConnections(currentConnection, currentHost);
+      return;
+    }
+
+    final HostInfo hostInfo = this.currentConnectionProvider.getCurrentHostInfo();
+    final String hostPattern = this.propertySet.getStringProperty(PropertyKey.clusterInstanceHostPattern).getValue();
+    final RdsHost rdsHost;
+
+    if (!StringUtils.isNullOrEmpty(hostPattern)) {
+      rdsHost = this.rdsHostUtils.getRdsHostFromHostPattern(this.propertySet, hostInfo, hostPattern);
+    } else {
+      rdsHost = this.rdsHostUtils.getRdsHost(this.propertySet, hostInfo);
+    }
+
+    final RdsUrlType rdsUrlType = rdsHost.getUrlType();
+    final String clusterId = rdsHost.getClusterId();
+    if (clusterId != null) {
+      this.topologyService.setClusterId(clusterId);
+    }
+    this.topologyService.setClusterInstanceTemplate(rdsHost.getClusterInstanceTemplate());
+
+    final List<HostInfo> topology = this.topologyService.getTopology(currentConnection, false);
+    if (Util.isNullOrEmpty(topology)) {
+      this.hosts.addAll(connectionUrl.getHostsList());
+      return;
+    }
+    this.hosts = topology;
+
+    if (StringUtils.isNullOrEmpty(hostPattern) &&
+        (IP_ADDRESS.equals(rdsUrlType) || OTHER.equals(rdsUrlType))) {
+      throw new SQLException(Messages.getString("ClusterAwareConnectionProxy.5"));
+    }
+
+    updateInternalConnections(currentConnection, topologyService.getHostByName(currentConnection));
+  }
+
+  private void updateInternalConnections(JdbcConnection currentConnection, HostInfo currentHost) {
+    if (currentHost == null) {
+      return;
+    }
+
+    final int currentHostIndex = this.rdsHostUtils.getHostIndex(this.hosts, currentHost.getHostPortPair());
+    if (currentHostIndex == 0) {
+      this.writerConnection = currentConnection;
+    } else if (currentHostIndex != -1) {
+      this.readerConnection = currentConnection;
+    }
```

Comment:
what happens if it's -1 ? should it ignore it or throw something?

Reply:
Good question. If its -1, that means the current connection does not match any of the hosts we have information about. I'm not sure if we would ever actually hit this case. If we do, my thoughts were that the current connection is probably still usable, so we can let the user use it for now. When the user calls setReadOnly, we will switch to a host that we actually have the information for. Let me know if you think an exception would be more suitable though